### ISSUE TITLE
[Truffle] String#rindex error from CGI specs

### ISSUE THREAD

**Issue by bjfish** [jruby/jruby#2802]
Example
```ruby
string = <<-eos
<HTML>
<BODY>
</BODY>
</HTML>


eos
element = "BODY"
string.rindex(/(?i-mx:^\s*<#{element})/i, 14)
```
Error
```
core:/core/rubinius/common/type.rb:559:in `initialize': undefined conversion for '"<HTML>\n<BODY>\n</BODY>\n</HTML>\n\n\n"' from UTF-8 to  (Encoding::CompatibilityError)
	from core:/core/rubinius/common/type.rb:559:in `raise'
	from core:/core/rubinius/common/type.rb:559:in `compatible_encoding'
	from core:/core/rubinius/common/string.rb:2317:in `rindex'
	from rindex.rb:10:in `<main>'
```

**Comment by chrisseaton** [comment]
Can I assign this to you, @nirvdrum, please?

**Comment by nirvdrum** [comment]
This looks more encoding related, but I'm happy to take a look.

**Comment by nirvdrum** [comment]
The problem here is our `Regexp` implementation and its very incomplete support for encodings and MBC encodings in particular.  Fixing this will require an overhaul of `Regexp`.

**Comment by bjfish** [comment]
Fixed at Fixed at 8bb41db63a78741bc0cd1de9b6eba2c7c5b869b6
